File: handlers/attitude_signal.py
"""
Attitude signal collector -- implicit attitude signals from hook events.

Collects two signal types:
1. Notification (tool_denied) -> autonomy_level correction
2. SessionEnd -> session statistics (deny count, message count, tool density)

Output: ~/Claude/memvault/corrections/auto/{YYYY-MM}/{date}.jsonl
"""

from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path

from .base import ALLOW, HOME, HookResult

CORRECTIONS_DIR = Path(HOME) / "Claude" / "memvault" / "corrections" / "auto"
SPOOL_DIR = Path(HOME) / ".hook-observatory" / "spool"
SPOOL_FILE = SPOOL_DIR / "events.jsonl"


def _write_correction(category: str, fact: str, session_id: str = "") -> None:
    """Append a correction record to the auto corrections JSONL."""
    now = datetime.now()
    month_dir = CORRECTIONS_DIR / now.strftime("%Y-%m")
    month_dir.mkdir(parents=True, exist_ok=True)

    out_file = month_dir / f"{now.strftime('%Y-%m-%d')}.jsonl"
    record = {
        "fact": fact,
        "category": category,
        "session_id": session_id,
        "timestamp": now.isoformat(timespec="seconds"),
        "source": "attitude_signal",
    }
    with open(out_file, "a", encoding="utf-8") as f:
        f.write(json.dumps(record, ensure_ascii=False) + "\n")


def _handle_notification(tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """Handle Notification events -- detect tool denials."""
    try:
        data = json.loads(raw_input) if isinstance(raw_input, str) else raw_input
    except (json.JSONDecodeError, TypeError):
        return ALLOW

    # Check for denial notification
    message = data.get("message", "") or ""
    notification_data = data.get("data", data)

    # Claude Code sends denial info in various formats
    denied_tool = ""
    if "denied" in message.lower():
        denied_tool = notification_data.get("tool_name", "") or tool_name
    elif notification_data.get("type") == "tool_denied":
        denied_tool = notification_data.get("tool_name", "")

    if denied_tool:
        session_id = notification_data.get("session_id", "")
        _write_correction(
            category="autonomy_level",
            fact=f"使用者 deny {denied_tool}，偏好更多確認再執行",
            session_id=session_id,
        )

    return ALLOW


def _handle_session_end(tool_input: dict, raw_input: str) -> HookResult:
    """Handle SessionEnd -- analyze spool statistics for the session."""
    try:
        data = json.loads(raw_input) if isinstance(raw_input, str) else raw_input
    except (json.JSONDecodeError, TypeError):
        return ALLOW

    session_id = data.get("data", {}).get("session_id", "") or data.get("session_id", "")
    if not session_id:
        return ALLOW

    # Read spool and filter for this session's events
    if not SPOOL_FILE.is_file():
        return ALLOW

    try:
        session_events = []
        with open(SPOOL_FILE, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    evt = json.loads(line)
                    evt_sid = evt.get("data", {}).get("session_id", "")
                    if evt_sid == session_id:
                        session_events.append(evt)
                except json.JSONDecodeError:
                    continue
    except Exception:
        return ALLOW

    if not session_events:
        return ALLOW

    # Count statistics
    deny_count = 0
    message_count = 0
    tool_count = 0

    for evt in session_events:
        event_type = evt.get("event_type", "")
        evt_data = evt.get("data", {})

        if event_type == "Notification":
            msg = evt_data.get("message", "") or ""
            if "denied" in msg.lower() or evt_data.get("type") == "tool_denied":
                deny_count += 1

        if event_type == "UserPromptSubmit":
            message_count += 1

        if event_type in ("PreToolUse", "PostToolUse"):
            tool_count += 1

    # Infer attitude signals
    if deny_count >= 3:
        _write_correction(
            category="autonomy_level",
            fact=f"本 session deny {deny_count} 次，使用者偏好更多確認",
            session_id=session_id,
        )

    if message_count > 0 and tool_count / message_count > 15:
        _write_correction(
            category="verbosity",
            fact=f"高工具密度 ({tool_count}/{message_count})，使用者可能偏好精簡對話",
            session_id=session_id,
        )

    return ALLOW


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """Main entry point -- dispatches by event_type."""
    if event_type == "Notification":
        return _handle_notification(tool_name, tool_input, raw_input)
    if event_type == "SessionEnd":
        return _handle_session_end(tool_input, raw_input)
    return ALLOW

File: internal/handlers/attitude_signal.go
package handlers

// attitude_signal.go — Go port of handlers/attitude_signal.py
//
// Collects implicit attitude signals from hook events:
//   - Notification (tool_denied)  → autonomy_level correction
//   - SessionEnd                  → session statistics from spool
//
// Output: ~/Claude/memvault/corrections/auto/{YYYY-MM}/{YYYY-MM-DD}.jsonl

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/joneshong/hook-dispatcher/internal/core"
)

func init() {
	entry := core.Entry{
		Matcher:    "",
		Handler:    attitudeSignalHandle,
		Critical:   false,
		ModuleName: "attitude_signal",
	}
	core.Register("Notification", entry)
	core.Register("SessionEnd", entry)
}

func attitudeSignalHandle(eventType, toolName string, _ map[string]any, rawInput string) core.HookResult {
	switch eventType {
	case "Notification":
		return asHandleNotification(toolName, rawInput)
	case "SessionEnd":
		return asHandleSessionEnd(rawInput)
	default:
		return core.Allow()
	}
}

// ---------------------------------------------------------------------------
// Notification: detect tool denials
// ---------------------------------------------------------------------------

func asHandleNotification(toolName, rawInput string) core.HookResult {
	var data map[string]any
	if rawInput != "" {
		if err := json.Unmarshal([]byte(rawInput), &data); err != nil {
			return core.Allow()
		}
	}

	message, _ := data["message"].(string)
	notifData, _ := data["data"].(map[string]any)
	if notifData == nil {
		notifData = data
	}

	deniedTool := ""
	if strings.Contains(strings.ToLower(message), "denied") {
		if v, _ := notifData["tool_name"].(string); v != "" {
			deniedTool = v
		} else {
			deniedTool = toolName
		}
	} else if t, _ := notifData["type"].(string); t == "tool_denied" {
		deniedTool, _ = notifData["tool_name"].(string)
	}

	if deniedTool != "" {
		sessionID, _ := notifData["session_id"].(string)
		asWriteCorrection(
			"autonomy_level",
			"使用者 deny "+deniedTool+", 偏好更多確認再執行",
			sessionID,
		)
	}

	return core.Allow()
}

// ---------------------------------------------------------------------------
// SessionEnd: analyze spool statistics for the session
// ---------------------------------------------------------------------------

func asHandleSessionEnd(rawInput string) core.HookResult {
	var data map[string]any
	if rawInput != "" {
		if err := json.Unmarshal([]byte(rawInput), &data); err != nil {
			return core.Allow()
		}
	}

	sessionID := ""
	if inner, ok := data["data"].(map[string]any); ok {
		sessionID, _ = inner["session_id"].(string)
	}
	if sessionID == "" {
		sessionID, _ = data["session_id"].(string)
	}
	if sessionID == "" {
		return core.Allow()
	}

	spoolFile := filepath.Join(core.Cfg().GetSpoolDir(), "events.jsonl")
	raw, err := os.ReadFile(spoolFile)
	if err != nil {
		return core.Allow()
	}

	var sessionEvents []map[string]any
	for _, line := range strings.Split(string(raw), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var evt map[string]any
		if err := json.Unmarshal([]byte(line), &evt); err != nil {
			continue
		}
		evtData, _ := evt["data"].(map[string]any)
		evtSID, _ := evtData["session_id"].(string)
		if evtSID == sessionID {
			sessionEvents = append(sessionEvents, evt)
		}
	}

	if len(sessionEvents) == 0 {
		return core.Allow()
	}

	denyCount := 0
	messageCount := 0
	toolCount := 0

	for _, evt := range sessionEvents {
		evtType, _ := evt["event_type"].(string)
		evtData, _ := evt["data"].(map[string]any)

		switch evtType {
		case "Notification":
			msg, _ := evtData["message"].(string)
			tp, _ := evtData["type"].(string)
			if strings.Contains(strings.ToLower(msg), "denied") || tp == "tool_denied" {
				denyCount++
			}
		case "UserPromptSubmit":
			messageCount++
		case "PreToolUse", "PostToolUse":
			toolCount++
		}
	}

	if denyCount >= 3 {
		asWriteCorrection(
			"autonomy_level",
			"本 session deny "+itoa(denyCount)+" 次, 使用者偏好更多確認",
			sessionID,
		)
	}

	if messageCount > 0 && float64(toolCount)/float64(messageCount) > 15.0 {
		asWriteCorrection(
			"verbosity",
			"高工具密度 ("+itoa(toolCount)+"/"+itoa(messageCount)+"), 使用者可能偏好精簡對話",
			sessionID,
		)
	}

	return core.Allow()
}

// ---------------------------------------------------------------------------
// Helper: append a correction record
// ---------------------------------------------------------------------------

func asWriteCorrection(category, fact, sessionID string) {
	home, _ := os.UserHomeDir()
	correctionsDir := filepath.Join(home, "Claude", "memvault", "corrections", "auto")

	now := time.Now()
	monthDir := filepath.Join(correctionsDir, now.Format("2006-01"))
	if err := os.MkdirAll(monthDir, 0o755); err != nil {
		return
	}

	outFile := filepath.Join(monthDir, now.Format("2006-01-02")+".jsonl")
	record := map[string]any{
		"fact":       fact,
		"category":   category,
		"session_id": sessionID,
		"timestamp":  now.Format("2006-01-02T15:04:05"),
		"source":     "attitude_signal",
	}
	line, err := json.Marshal(record)
	if err != nil {
		return
	}

	f, err := os.OpenFile(outFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return
	}
	defer f.Close()
	f.Write(line)
	f.Write([]byte("\n"))
}
